### PR TITLE
159 osmvalidate module

### LINKED ISSUE #159: osm.validate module
**Description of the Feature to be Added**

A class instance used for validating the OSM files produced in filtering with osmosis. This may have simple summary tables, such as counts of tags by tag type. Lists of unique tag IDs. 

Returning the tag hierarchy for each tag would be a nice to have.

Optionally methods to visualise the features against a base map would be useful. Either using osmnx or reverse engineering the functionality we need from pyrosm.

**(OPTIONAL) Suggested Implementations**

**Additional context**

Help QA filtering strategies.
Help QA varying OSM quality in different territories.

### LINKED ISSUE #194: Visualise OSM spatial features on basemap
A 'nice to have' element of issue https://github.com/datasciencecampus/transport-network-performance/issues/159 was to visualise spatial features of interest against a base map. This is doable, though will take more time. Therefore if the reviewer agrees, I will nudge that request over into a separate issue.

_Originally posted by @r-leyshon in https://github.com/datasciencecampus/transport-network-performance/issues/193#issuecomment-1782505839_

### PR DESCRIPTION
**Pull request by r-leyshon**
## Description

Fixes #159 Fixes #194

## Motivation and Context

Adds some basic functionality for checking spatial features in an OSM file, such as tags, locations and IDs.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Previously there was an issue with skipping expensive tests (Please see #189). The updated approach avoids the need to skip or mark, by filtering the osm.pbf fixture to a very small bounding box, test durations become trivial.

Test configuration details:
* OS: macos
* Python version: 3.9.13
* Java version: openjdk 11.0.19 2023-04-18 LTS
* Python management system: pip

## Advice for reviewer

Please see note in requirements.txt & contributing.md. I had some trouble building an environment for pyosmium on the new m1 mac. Turns out I was missing a few brew installs that were already present on my old machine. 

## Checklist:

- [X] My code follows the intended structure of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

## Additional comments

A 'nice to have' element of issue #159 was to visualise spatial features of interest against a base map. This is doable, though will take more time. This feature was previously expected to have been part of a separate PR but is now included along with the rest of the osm validation logic.

### CODE REVIEW

**Comment by r-leyshon** [comment]
A 'nice to have' element of issue https://github.com/datasciencecampus/transport-network-performance/issues/159 was to visualise spatial features of interest against a base map. This is doable, though will take more time. Therefore if the reviewer agrees, I will nudge that request over into a separate issue.

File: src/transport_performance/osm/validate_osm.py
Comment (SergioRec) [inline src/transport_performance/osm/validate_osm.py]: This function can be made more concise:

```
def _compile_tags(osmium_feature):
    """docstring"""
    tagdict = {tag.k: tag.v for tag in osmium_feature.tags}
    return tagdict
```

Idea from [here](https://github.com/osmcode/pyosmium/issues/106).

Current test for the function won't work with this though. It should be possible to create a `osmium_feature.tags` fixture from the loaded pbf file.

File: src/transport_performance/osm/validate_osm.py
Comment (SergioRec) [inline src/transport_performance/osm/validate_osm.py:12]: Could we add something about what the module actually does? As in "gets lists of nodes, ways, etc. with ids, tags and coordinates", perhaps an example of outputs. The osmium module is quite obscure, same with the actual contents of a pfb file. I think this would help users greatly (it would definitely help me!).

Since the module is called `validate_osm`, how do we use tthis module to validate them? Do we plan to check specific elements, filter some elements out, or replace information that we know might be wrong?

Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py:12]: Thanks for this, I have reviewed the language and added a brief list of the functionality currently offered.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: @SergioRec this is a great find! No worries on the test fixture not working with this approach as the fixture was only a simple representation of a pyosmium class, it didn't have all the necessary attributes to be able to work with the dictionary comprehension.
The good news is that the approach you found made the tests for _compile_tags defunct. As I had made assertions about tag contents further down the test module, these have all passed.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: This is a small thing, but it would be good to import the order correctly in compliance with PEP 8. [See here ](https://peps.python.org/pep-0008/#imports).  

**Suggested Implementation (Optional)**  
```python
# standard library imports
from pathlib import Path
from typing import Union

# third party importrs
import osmium
import pandas as pd
import geopandas as gpd
from shapely import Point
import folium

# local imports
```

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: This docstring refers to `tagdict` as a return, however the function does not return a named variable.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: In terms of docstrings here, I believe each parameter should be formatted with a space after the parameter name. Also, there is no need for parenthesis around the types.  

For example   
`osm_dict :  dict` rather than `osm_dict: (dict)`.  

Raising this as I'm not sure how Sphinx reacts to this change in formatting.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: Reading through this logic here, it appears that `mem_row` is being re-defined before usage, making the first definition useless.  
It also appears that while iterating through the members, only the last member `(mem_row)` is being concatted to `out_df`.  
To describe this better, the for loop `for k, j in i.items()` is constantly redefining `mem_row,` without adding it to out_df``

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: I assume this is to be removed and was left in after debugging.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py:385]: Type hinting for `w`

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py:371]: Type hinting for `n`

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py:40]: Type hinting for  `osm_feature`

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: This function takes `accepted_keys` as a param. I am torn between keeping this as a param and removing it.  

It seems like an inconvenience to have as a required parameter, as in most instances of it's implementation, you would expect all dict keys to be accepted.

In cases where they aren't, we could defend aginst that outside of the function, rather than through a parameter.

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py:129]: An assertion for `len(out) == 3` could be useful here.  
This would ensure that we haven't removed any of the keys with the value `"keep"`.

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py]: This assertion message should be `f"Expected latitude value of 2.0, but found {exp_lat}"`.  
Small typo

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py]: Small typos here. Should be 2.0 instead of 1.0

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py:536]: Is it worth asserting that the index is correct as the outputted GDF uses a  `MultiIndex`.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: Is this an easier implementation?. It may not be preferred however as it causes `.count_features()` to become dependent on `.get_feature_ids()`.  
```python
{i: len(j) for i, j in test.get_feature_ids().items()}
```

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py]: Should we also check for methods inherited from `osmium.SimpleHandler` ?  
An example of this would be `.apply_file()`

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: I feel that this docstring may be a bit unclear. I understand that its describing the structure of the returned dictionary, however this wasn't clear at first. An example may help the user in this instnce.

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py:1]: Is it possible to add a small test for `_compile_tags`?. While this is a very small function, it may be useful to have.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: Would this implementation be more suitable. Functionality is the same however this seems more dynamic for adding other features.  

```python
if feature_type not in ["node", "way"]:
    raise NotImplementedError(
       "Relation or area plotting not implemented at this time."
    )
```

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [reply tests/osm/test_validate_osm.py]: This also applies to similar classes such as `FindLocations`.

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py:158]: It may be useful to add a `test_findids_raises` section where we test that the correct errors are raised when an invalid path is passed.  
This would also apply to similar classes.

File: tests/osm/test_validate_osm.py
Comment (cbrowning04) [inline tests/osm/test_validate_osm.py]: I'm not sure what your opinion on this is, however it may be useful to use `folium.Map` or a name such as `folium_map` to avoid confusion with the built in `map`.

File: pipeline/osm/01-validate-osm.py
Comment (cbrowning04) [inline pipeline/osm/01-validate-osm.py]: Very picky but are able we to start the sentence with a capital? 🙃 😆

File: pipeline/osm/01-validate-osm.py
Comment (cbrowning04) [inline pipeline/osm/01-validate-osm.py]: Same as above  😆

**Review by cbrowning04 — CHANGES_REQUESTED** [review]
Hi @r-leyshon. Great work on this. Good job integrating with pyosmium under the limitations of the package,  a very neat implementation overall.  

It may seem like a lot of comments, but a lot of them re docstring related/very small accuracy changes so nothing to worry about.

The overall functionality of the module is great. Everything working as expected.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [reply src/transport_performance/osm/validate_osm.py]: ID  : {name:value}

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Good point. @ethan-moss we should review our pre-commit setup. I though isort does this for us.

Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: done

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Good catch, this was inconsistent throughout. Done.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Good spot, sorry! Done.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py:385]: Rather not type hint the internals if that's acceptable. I wouldn't expect people to be working with them if they didn't understand it.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py:371]: as above

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py:40]: as above

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py:129]: Done

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py]: good find. done.

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py]: Done.

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py]: No longer relevant as inheritance removed and name mangled attributes introduced.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py]: Done

File: pipeline/osm/01-validate-osm.py
Comment (r-leyshon) [reply pipeline/osm/01-validate-osm.py]: Done

File: pipeline/osm/01-validate-osm.py
Comment (r-leyshon) [reply pipeline/osm/01-validate-osm.py]: Done

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: This was a really good spot and well described. Thanks @CBROWN-ONS I've put this right now and updated the tests to account for expected number of rows in GDF output (1 row for each node member in a way).

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Agreed. On reviewing the purposes I felt this could be tidied up a lot. New approach ditches the accepted_keys parameter altogether and implements a utility method for FindLocations that will check the user is asking for anything other than node or way in the methods and raise a NotImplementedError.

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py:536]: Done

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Agreed Done.

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py:1]: OK with the coverage right now as this would mean mocking osmium internals. If you feels strongly that this should be independently covered then happy to raise as a tech debt issue.

File: tests/osm/test_validate_osm.py
Comment (r-leyshon) [reply tests/osm/test_validate_osm.py:158]: Done

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: This should be `tag_collator.` Currently `tag_collator` is unused.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: This should be `loc_collator.` Currently `loc_collator` is unused.

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: It would be useful to have a type defence for `id_collator` to ensure that it is an instance of `_IdHandler.` The same applies to `FindLocations` and  `FindTags`

File: src/transport_performance/osm/validate_osm.py
Comment (cbrowning04) [inline src/transport_performance/osm/validate_osm.py]: Type in the docstring here. `=_LocHandler` can be removed and replaced with `loc_collator : _LocHandler, optional`

**Review by cbrowning04 — CHANGES_REQUESTED** [review]
Hi @r-leyshon. Great job clearing up those other comments so fast.  I've got a few quick fixes on the new stuff that will need changing before merging, but once they're done  I'm happy to merge

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Good spot. Sorry - clearly rushing. I've implemented this fix.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done. Tested.

File: src/transport_performance/osm/validate_osm.py
Comment (r-leyshon) [reply src/transport_performance/osm/validate_osm.py]: Done.

**Comment by r-leyshon** [comment]
Thanks @CBROWN-ONS , those should all be done now.

**Review by cbrowning04 — APPROVED** [review]

**Pull request merged**
